rangefeed: drop messages for overflowed streams

stevendanna opened the pull request:
This PR fixes some bugs discovered when trying to increase the test coverage of the buffered sender unit tests:

- A stream would actually never move into the "overflowing" state. Rather it would stay in the "active" state and keep returning a buffer capacity error until the registration sent the error back to us, in which case it would go straight to overflowed.

- This was papering over another bug in which the overflowing state could produce an invalid stream of events.

Here, I add test coverage for the expected state transitions and fix these bugs by:

  - Dropping buffered messages for streams not in the stream map. This works because the stream will be in the stream map from the point of the RegisteringStream call and is only removed when the stream finally sends the error. Thus any message for a stream not in map represents either (1) a message sent to a StreamID before RegisteringStream has been called or (2) a message sent to a StreamID after an Error. (1) would be a programming error and (2) would be messages that are dropped server-side anyway.

  - Re-organizing the code such that it is harder to miss saving the stream status back to the status map.

Note any potential bug here only existed on master and thus no release note is needed.

Epic: CRDB-55217

Release note: None

stevendanna committed:
rangefeed: drop messages for overflowed streams

This PR fixes some bugs discovered when trying to increase the test
coverage of the buffered sender unit tests:

- A stream would actually never move into the "overflowing" state.
  Rather it would stay in the "active" state and keep returning a buffer
  capacity error until the registration sent the error back to us, in
  which case it would go straight to overflowed.

- This was papering over another bug in which the overflowing state
  could produce an invalid stream of events.

Here, I add test coverage for the expected state transitions and fix
these bugs by:

  - Dropping buffered messages for streams not in the stream map. This
  works because the stream will be in the stream map from the point of
  the RegisteringStream call and is only removed when the stream finally
  sends the error. Thus any message for a stream not in map represents
  either (1) a message sent to a StreamID before RegisteringStream has
  been called or (2) a message sent to a StreamID after an Error. (1)
  would be a programming error and (2) would be messages that are
  dropped server-side anyway.

  - Re-organizing the code such that it is harder to miss saving the
  stream status back to the status map.

Note any potential bug here only existed on master and thus no release
note is needed.

Epic: CRDB-55217

Release note: None

stevendanna committed:
rangefeed: reject all buffered sends after error

This changes the flow control in the buffered sender to allow us to
immediately start dropping buffered sends after the first error is
enqueued. This simplifies the code, makes it easier to write assertions
about message ordering, and avoids cases where we have unaccounted for
memory used by messages that were in the queue after an error.

Epic: CRDB-55217
Release note: None

stevendanna committed:
rangefeed: deflake TestBufferedSenderDisconnectStream under stress

This uses unique stream IDs per test to avoid having to coordinate test
cases. Further, we now explicitly wait for error events rather than for
buffer empty events to avoid concurrency issues where we observe the
buffer being empty before it was sent to the test stream.

Epic: none
Release note: None

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-:-OclYYMm00j9TiNmIJqk:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`-- commits` line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-OclXH0-2-EzU0zGs0VF:-OclXH0-2-EzU0zGs0VG:b-s23cn3):*
Is this only happening due to `publishCatchUpBuffer`as well (live event path should check for disconnected under lock?) or is there somewhere else 
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-OclB4ir7IiBpVis9UYz:-OclB4ir7IiBpVis9UZ-:bi4k1tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fab29290a908b22bbd3199dc2bd972e69d39a8/pkg/kv/kvserver/rangefeed/buffered_sender.go#L161)):*
> ```Go
> // stream. It does not block. It errors in the case of a stopped sender of if
> // the registration has exceeded its capacity.
> func (bs *BufferedSender) sendBuffered(
> ```

nit: we should clarify that no error is returned when already disconnected or already overflowing, and we rely on the caller to release one usage on alloc
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 206 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-OclAkusEDCpj-OPKzV-:-OclAkusEDCpj-OPKzV0:bfd31ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fab29290a908b22bbd3199dc2bd972e69d39a8/pkg/kv/kvserver/rangefeed/buffered_sender.go#L206)):*
> ```Go
> func (bs *BufferedSender) nextPerQueueStateLocked(
> 	status streamStatus, ev *kvpb.MuxRangeFeedEvent,
> ) (streamState, bool, error) {
> ```

nit: should we use named return values here to make it clear that these represent the next state and shouldAdmit


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
TFTR!  I'll push some comment updates and merge in a bit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-:-Ocm9Eej6jKGmTXWaO6E:b-a9hgjt)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-OclXH0-2-EzU0zGs0VF:-Ocm8N598tx9D-Qd0cUC:buvxbl9):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Is this only happening due to `publishCatchUpBuffer`as well (live event path should check for disconnected under lock?) or is there somewhere else
</blockquote></details>

Yeah, basically all of the bugs are only possible because of publishCatchUpBuffer.
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-OclB4ir7IiBpVis9UYz:-Ocm8l9rBAO8hWgN5d5o:b-xgch98) ([raw file](https://github.com/cockroachdb/cockroach/blob/63fab29290a908b22bbd3199dc2bd972e69d39a8/pkg/kv/kvserver/rangefeed/buffered_sender.go#L161)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: we should clarify that no error is returned when already disconnected or already overflowing, and we rely on the caller to release one usage on alloc
</blockquote></details>

We had a lot of these comments previously, and am not sure they actually make things clearer. The caller should be releasing the alloc regardless of what we do with the item. It's our responsibility to call Use() if we retain the value, its the caller's responsibility to call Release() if they don't retain the value. That is the same contract everywhere.


<!-- Sent from Reviewable.io -->

stevendanna committed:
rangefeed: return error from BufferedSender on dropped events

The API of sendBuffered is a bit clearer if it always returns an error
if the event isn't admitted to the buffer.

Epic: none
Release note: None

stevendanna committed:
rangefeed: cleanup test output

A couple more cases of missing log.Scopes and some uses of fmt.Printf
rather than t.Log.

Epic: none
Release note: None

stevendanna commented:
The AI reviewer pointed out something that _was_ purposeful because the missaccounting was inconsequential and I liked the ability to remove another branch. But, I suppose in retrospect it might have confused some humans too so I've given it a point.

stevendanna committed:
rangefeed: rename nextPerQueueStateLocked to nextPerStreamStateLocked

Epic: none
Release note: None

stevendanna committed:
rangefeed: dedup TestBufferedSenderDisconnectStream

This tests duplicated TestStreamManagerDisconnectStream, which now
supports both sender types.

Epic: none
Release note: None

arulajmani reviewed (approved):
I've only looked at the last few commits that Wenyi did not get a chance to. 

> @arulajmani reviewed 2 of 4 files at r1, 8 of 8 files at r10, 1 of 1 files at r11, 4 of 4 files at r12, 2 of 2 files at r13, 3 of 3 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156458#-:-OcvpKclAiC5vPqyrY-W:b-rt91xt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna and @wenyihu6)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors retry
